vllm/config/utils.py

```
@@ -18,12 +18,137 @@ def config(cls: ConfigT) -> ConfigT:
+    argument provided by `get_kwargs` will be
     `pydantic.TypeAdapter(ConfigT).validate_json(cli_arg)` which treats the
     `cli_arg` as a JSON string which gets validated by `pydantic`.
 
     Config validation is performed by the tools/validate_config.py
     script, which is invoked during the pre-commit checks.
     """
     return cls
+
+
+# Canonicalize common value types used in config hashing.
+# Keeps imports light at module load; heavier bits are imported inside.
+def canon_value(x):
+    """Return a stable, JSON-serializable canonical form for hashing.
+
+    Fast path primitives first, then special types (Enum, callable, torch dtype
+    and Path), then generic containers (Mapping/Set/Sequence) with recursion.
+    """
+    import enum
+    import pathlib
+    from collections.abc import Mapping, Sequence, Set
+
+    # Fast path
+    if x is None or isinstance(x, (bool, int, float, str)):
+        return x
+
+    # Enums by value
+    if isinstance(x, enum.Enum):
+        return x.value
+
+    # Callable by qualified name
+    try:
+        if callable(x):
+            module = getattr(x, "__module__", "")
+            qual = getattr(x, "__qualname__", repr(x))
+            return f"{module}.{qual}" if module else qual
+    except Exception:
+        pass
+
+    # Torch dtype without hard dependency
+    try:
+        import torch  # noqa: WPS433 (local import by design)
+        if isinstance(x, torch.dtype):
+            return str(x)
+    except Exception:
+        pass
+
+    # Paths
+    if isinstance(x, pathlib.Path):
+        return str(x)
+
+    # Containers (generic)
+    if isinstance(x, Mapping):
+        return tuple(sorted((str(k), canon_value(v)) for k, v in x.items()))
+    if isinstance(x, Set):
+        return tuple(sorted(repr(canon_value(v)) for v in x))
+    if isinstance(x, Sequence) and not isinstance(x, (str, bytes, bytearray)):
+        return tuple(canon_value(v) for v in x)
+
+    # Unsupported type
```

Can we have some kind of debug logging for unsupported types?  It seems like if an unsupported member type is encountered, it silently skips including it in the hash?  I feel like this could lead to subtle/hard to find bugs.

---

Fixed in most recent commit